### PR TITLE
Fetchr integration

### PR DESCRIPTION
**Pull request by geekyme**
releasing this early so you can see what i've made.

What I've completed: 
- an abstraction over superagent with promises added (works like fetchr) called ApiClient
- a middleware which works with the above and dispatch *_REQUEST, *_SUCCESS, *_FAILURE actions
- an action creator `getPhoto` that uses ApiClient and works with the middleware

Not completed:
- reducer to receive the actions from `getPhoto` and update the store
- connect the PhotoPage to rerender on store change
- tests
- readme

### CODE REVIEW

**Comment by gpbl** [comment]
Thanks @geekyme for your brave exploration, it is an invaluable help for me to understand which problems we are going to solve 👍 

So I'm wrapping my head around `ApiClient`: we should find a better naming because `client` is confusing. 

I try to summarize what you are doing, please correct me if I'm wrong.

* there's a middleware for Redux intercepting a `promise` key in the action creator
* `promise` is a function to make async calls to the API server. The action is not dispatched until the call is finished
* `promise` receives as argument an instance of the `ApiClient` utility (called `client`) that is a wrapper to superagent
* `promise` must return a `Promise`, which is resolved with the body of the superagent response, and rejected with the body *or* the response error. As the promise is fulfilled, the middleware hooks one of the 3 different types of this kind of action: `SUCCESS`, `REQUEST` or `FAILURE`.

I fear that such architecture is too much opinionated and a bit hard to grasp for newbies. If we want to describe this behavior in the README i think it will not so easy. It also may cut out some use cases that we can't foresee now: e.g., apiClient's promises should be fulfilled with the whole `req` object, not just with its body. 

That said, I would evaluate the use of Fetchr. ApiClient currently is not doing what fetchr does: fetchr creates a middleware in express that will intercept calls to a specific endpoint, e.g. `/api`, and it gives the freedom to implement a "service" that can be used for everything that is asynchronous *and* requires the express server in between: requests to an API server (with the library you prefer), but also to any server-side resource (think, for example, reading a file stored on the server). 

In fact, the code which is now in the `promise` key in `action/photo.js` should run only on the server, never on the client: this is how fetchr makes sure the API requests are "universal".
I know fetchr asks the developer to make some opinionated choices, i.e. the `CRUD` interface, but using fetchr we would shift this concern to an external library and not assume the developer adopts some conventions used only for redux500. 

...and there's also space for a redux-fetchr-middleware to be published on npm :P

File: src/actions/photo.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { PHOTO_LOAD_REQUEST, PHOTO_LOAD_SUCCESS, PHOTO_LOAD_FAILURE } from "../constants/ActionTypes";
+
+export function getPhoto(id) {
+  return {
+    types: [
+      PHOTO_LOAD_REQUEST,
+      PHOTO_LOAD_SUCCESS,
+      PHOTO_LOAD_FAILURE
+    ],
```
Comment (gpbl) [inline src/actions/photo.js]: Keep in mind I didn't read anything yet about redux middlewares: but is an array here a redux convention? Couldn't we write a object: 

```
types: {
   onRequest: PHOTO_LOAD_REQUEST,
   onSuccess: PHOTO_LOAD_SUCCESS,
   onFailure: PHOTO_LOAD_FAILURE
}
```

**Comment by gpbl** [comment]
Also, very important to follow: Flux Compliant Actions guidelines :-)
https://github.com/acdlite/flux-standard-action

**Comment by gpbl** [comment]
...and maybe adopt: https://github.com/acdlite/redux-actions

**Comment by geekyme** [comment]
I think we should avoid using Fetchr as much as possible because of 2 reasons:

1. Fetchr is too opinionated on the CRUD. This makes sense if we are usually handling CRUD .a.k.a. making database calls within each service. We are not doing that, as such we will simply end up with many files that all end up doing the same thing: proxying a request to the external API. 

2. Fetchr and its matrix URI schemes. It can create some bugs for instance a variable to be passed to the  backend for evaluation it sends that information like this `;collectionId=${collectionId}`. When collectionId is defined, it works well. However once it is undefined, it ends up sending this `;collectionId=undefined`, my backend read collectionId as the "undefined" string instead of the type undefined. 


I do agree that ApiClient may be a little hard to understand. It's a solution that comes without the above problems, but retain the spirit of what we want to do: proxy our api requests to the external API (500px) to avoid cross-domain issues. However, its still not perfect and we will have to find a way to hook ApiClient to express so that: 

1. Server-side rendered components that call methods on ApiClient directly, instead of issuing an XHR to itself. Kind of like how Fetchr makes internal calls on the server side. 

2. Client-side will make an XHR request to ApiClient via an express middleware with the prefix `/api`

> Keep in mind I didn't read anything yet about redux middlewares: but is an array here a redux convention? Couldn't we write a object:

Yes we can. I was just using array destructuring. Didn't know about the Flux Compliant Actions. I'll take a look. Thanks!

**Comment by geekyme** [comment]
hey @gpbl try this out, i have hooked up the reducers and connectors but there seem to be a bug. 

fetchComponentData is called only on the HomePage :S
also it seems like server side data is not being dehydrated to the client side properly. 

on the brightside fetchr is working well

**Comment by geekyme** [comment]
oh crap ok on the server side the data is not being resolved because `fetchrMiddleware` is not returning a promise. Changing it back to a promise

**Comment by geekyme** [comment]
ok fixed. now the only pesky issue is why fetchData is being ignored if i defined it on `PhotoPage`... and will only work if I defined it on `HomePage`.

**Comment by geekyme** [comment]
The promise is required because of the transition hook `fetchComponentData` i think, which in turns will grab the `fetchData` methods of the component class and fire them. Again this whole `fetchData` cruft will have to be revisited in a new branch. Have documented the issues surround it and react-router.

**Comment by gpbl** [comment]
I was wondering because AFAIK redux middlewares don't need to return promises to work. Weird that the `fetchComponentData` was working with them. But well, let see first what we do with the router :-)

**Pull request merged**